crates/modelardb_server/src/storage/types.rs

```diff
@@ -55,70 +65,83 @@ impl MemoryPool {
     ) -> Self {
         // unwrap() is safe as i64::MAX is 8192 PiB and the value is from ConfigurationManager.
         Self {
-            remaining_uncompressed_memory_in_bytes: RwLock::new(
+            wait_for_uncompressed_memory: Condvar::new(),
+            remaining_uncompressed_memory_in_bytes: Mutex::new(
                 uncompressed_memory_in_bytes.try_into().unwrap(),
             ),
-            remaining_compressed_memory_in_bytes: RwLock::new(
+            remaining_compressed_memory_in_bytes: Mutex::new(
                 compressed_memory_in_bytes.try_into().unwrap(),
             ),
         }
     }
 
     /// Change the amount of memory available for uncompressed data by `size_in_bytes`.
     pub(super) fn adjust_uncompressed_memory(&self, size_in_bytes: isize) {
-        // unwrap() is safe as write() only returns an error if the lock is poisoned.
-        *self.remaining_uncompressed_memory_in_bytes.write().unwrap() += size_in_bytes
+        // unwrap() is safe as lock() only returns an error if the mutex is poisoned.
+        *self.remaining_uncompressed_memory_in_bytes.lock().unwrap() += size_in_bytes;
+        self.wait_for_uncompressed_memory.notify_all();
     }
 
     /// Return the amount of memory available for uncompressed data in bytes.
     pub(super) fn remaining_uncompressed_memory_in_bytes(&self) -> isize {
-        // unwrap() is safe as write() only returns an error if the lock is poisoned.
-        *self.remaining_uncompressed_memory_in_bytes.read().unwrap()
+        // unwrap() is safe as lock() only returns an error if the mutex is poisoned.
+        *self.remaining_uncompressed_memory_in_bytes.lock().unwrap()
     }
 
-    /// Try to reserve `size_in_bytes` bytes of memory for uncompressed data. Returns [`true`] if
-    /// the reservation succeeds and [`false`] otherwise.
-    #[must_use]
-    pub(super) fn try_reserve_uncompressed_memory(&self, size_in_bytes: usize) -> bool {
-        // unwrap() is safe as write() only returns an error if the lock is poisoned.
-        let mut remaining_uncompressed_memory_in_bytes =
-            self.remaining_uncompressed_memory_in_bytes.write().unwrap();
-
-        let size_in_bytes = size_in_bytes as isize;
+    /// Reserve `size_in_bytes` bytes of memory for uncompressed data. For simplicity and to not
```

Comment:
> Seems odd to talk about how spilling works here since this function is not directly related to spilling buffers to disk. Maybe find a more relevant place to mention this. Also i am not sure i quite understand the comment, it is my understanding that we only spill finished (full) data buffers now.

I have tried to simplify the comment and make it more clear.

> Also this might not be relevant to change in this PR but since we in multiple places have seen the reserved memory as "soft" limits that can be passed momentarily, it might be relevant to mention this in the documentation to the user to make sure they do not expect that these are hard limits.

I have added a task to do this.